Parse captured output by chunks to save memory from large outputs

remss opened the pull request:
When a run prints lots of data to stdout then sacred takes more and more memory and time to capture the output. If you run an experience with a Keras progress bar, an epoch will take more and more time to process .

Here a sample example to show what I'm talking about:

```
import sacred
import time
import hashlib
import psutil
import os

ex = sacred.Experiment('issue')
ex.captured_out_filter = sacred.utils.apply_backspaces_and_linefeeds

@ex.automain
def main(_seed, _run):
    p = psutil.Process(os.getpid())
    for x in range(200):
        start = time.monotonic()
        for i in range(300000):
            rnd_val = str(i + time.monotonic()).encode('utf-8')
            rnd_val_hash = hashlib.sha256(rnd_val).hexdigest()
            if i % 1000:
                print('%06d' % i, rnd_val_hash, end="\r")
        duration = time.monotonic() - start
        log = 'duration = %05.2f sec | mem = % 4d MB'% (
            duration,
            p.memory_info().rss // 1024 // 1024
        )
        print(log.ljust(100))
```

Here is the memory usage and iteraton durations:

![sacred_mem_issue2](https://cloud.githubusercontent.com/assets/1586257/24789612/18a7dfba-1b74-11e7-9614-6b888bce9e6d.png)

I made a a patch to read stdout file descriptor from the last read position to avoid memory issue.

Memory usage and iteraton durations after the patch:

![sacred_mem_issue_patched2](https://cloud.githubusercontent.com/assets/1586257/24789618/1f6b86d0-1b74-11e7-8bd4-51cd0f9825d2.png)


Note: Even if the patch is for v0.7, this issue is present in v 0.6 too.

Qwlouse commented:
You are right, that is a problem indeed. Thanks for tackling this! 
But it seems that your change broke a few tests. Could you run `pytest tests/test_utils.py` and ensure that they complete successfully?

remss commented:
Fixed here and there. Should be OK now.
FYI I changed the tests suite to except final \r from apply_backspaces_and_linefeeds result because I need it to concat with previous output chunk. And the tee_output functions no longer tries to read and return the final_out since it can be too large.

Qwlouse commented:
It looks better now, thanks! And sorry for not reacting yet. I've been really busy and will be travelling this week. But I'll take a final look and merge it next week.

Qwlouse commented:
Thanks a lot!
